pennylane_lightning/core/src/simulators/lightning_gpu/measurements/MeasurementsGPU.hpp

```diff
@@ -0,0 +1,790 @@
+
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+/**
+ * @file
+ * Defines a class for the measurement of observables in quantum states
+ * represented by a Lightning Qubit StateVector class.
+ */
+
+#pragma once
+
+#include <algorithm>
+#include <complex>
+#include <cuda.h>
+#include <cusparse.h>
+#include <custatevec.h> // custatevecApplyMatrix
+#include <random>
+#include <type_traits>
+#include <unordered_map>
+#include <vector>
+
+#include "MeasurementsBase.hpp"
+#include "Observables.hpp"
+#include "ObservablesGPU.hpp"
+#include "StateVectorCudaManaged.hpp"
+#include "cuda_helpers.hpp"
+
+/// @cond DEV
+namespace {
+using namespace Pennylane;
+using namespace Pennylane::Measures;
+using namespace Pennylane::Observables;
+using namespace Pennylane::LightningGPU::Observables;
+namespace cuUtil = Pennylane::LightningGPU::Util;
+using Pennylane::LightningGPU::StateVectorCudaManaged;
+using namespace Pennylane::Util;
+} // namespace
+/// @endcond
+
+namespace Pennylane::LightningGPU::Measures {
+/**
+ * @brief Observable's Measurement Class.
+ *
+ * This class couples with a statevector to performs measurements.
+ * Observables are defined by its operator(matrix), the observable class,
+ * or through a string-based function dispatch.
+ *
+ * @tparam StateVectorT type of the statevector to be measured.
+ */
+template <class StateVectorT>
+class Measurements final
+    : public MeasurementsBase<StateVectorT, Measurements<StateVectorT>> {
+  private:
+    using PrecisionT = typename StateVectorT::PrecisionT;
+    using ComplexT = typename StateVectorT::ComplexT;
+    using BaseType = MeasurementsBase<StateVectorT, Measurements<StateVectorT>>;
+    using CFP_t = decltype(cuUtil::getCudaType(PrecisionT{}));
+    cudaDataType_t data_type_;
+
+    GateCache<PrecisionT> gate_cache_;
+
+  public:
+    explicit Measurements(const StateVectorT &statevector)
+        : BaseType{statevector},
+          gate_cache_(true, statevector.getDataBuffer().getDevTag()) {
+        if constexpr (std::is_same_v<CFP_t, cuDoubleComplex> ||
+                      std::is_same_v<CFP_t, double2>) {
+            data_type_ = CUDA_C_64F;
+        } else {
+            data_type_ = CUDA_C_32F;
+        }
+    };
+
+    /**
+     * @brief Utility method for probability calculation using given wires.
+     *
+     * @param wires List of wires to return probabilities for in lexicographical
+     * order.
+     * @return std::vector<PrecisionT>
+     */
+    auto probs(const std::vector<size_t> &wires) -> std::vector<PrecisionT> {
+        // Data return type fixed as double in custatevec function call
+        std::vector<double> probabilities(Pennylane::Util::exp2(wires.size()));
```

Comment:
This still needs to be double, rather than float/double?

Reply:
yes, it should be `double`.